napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -58,42 +58,27 @@ def _rebuild_npe1_samples_menu() -> None:  # pragma: no cover
     if unreg := plugin_manager._unreg_sample_actions:
         unreg()
 
-    # sample_actions: List[Action] = []
+    sample_actions: List[Action] = []
+    sample_submenus: List[Any] = []
     for plugin_name, samples in plugin_manager._sample_data.items():
         multiprovider = len(samples) > 1
         if multiprovider:
             submenu_id = f'napari/file/samples/{plugin_name}'
-            submenu = [
-                (
-                    MenuId.FILE_SAMPLES,
-                    SubmenuItem(
-                        submenu=submenu_id, title=trans._(plugin_name)
-                    ),
-                ),
-            ]
+            submenu = (
+                MenuId.FILE_SAMPLES,
+                SubmenuItem(submenu=submenu_id, title=trans._(plugin_name)),
+            )
+            sample_submenus.append(submenu)
         else:
             submenu_id = MenuId.FILE_SAMPLES
-            submenu = []
-        sample_actions: List[Action] = []
+
         for sample_name, sample_dict in samples.items():
 
-            def _add_sample(
-                qt_viewer: QtViewer,
-                plugin: str = plugin_name,
-                sample: str = sample_name,
-            ) -> None:
-                from napari._qt.dialogs.qt_reader_dialog import (
-                    handle_gui_reading,
-                )
-
-                try:
-                    qt_viewer.viewer.open_sample(plugin, sample)
-                except MultipleReaderError as e:
-                    handle_gui_reading(
-                        [str(p) for p in e.paths],
-                        qt_viewer,
-                        stack=False,
-                    )
+            _add_sample_partial = partial(
+                _add_sample,
```

Comment:
Where add sample is defined? It is function used currently by `npe2`?

Reply:
Yeah, the callback is the same for npe1 and npe2 for samples